Use colors to signal bad sampling

PhilipVinc opened the pull request:
This was a little weekend's experimentation. 

I don't have the time to finish this now but would love some feedback on it...

The idea is to prevent users to completely ignore when sampling is going bad. We currently have 2 metrics: `variational_state.sampler_state.acceptance` (which is displayed in the repr of variational state) and `R_hat` of the energy.

I tried using [rich](https://github.com/Textualize/rich#readme) to interpolate between green (`R_hat==1`)  and red (`(1-R_hat)>>0`). Similar coloring is used for `sampler_state.acceptance`.
The color curve is a sigmoid, and was just chosen pretty much arbitrarily.  I chose `rich` because it's the only python package that supports arbitrary colors (and not only 16 colors) that i could find. 
This works in python repl and in jupyter notebooks. 

Interestingly, rich also has advanced support for progress bars with crazy displays so we could use it to also display additional data during training. In this PR there is an experimental progress bar with rich.

You can see the result here (the progress bar normally displays correctly and not on a new line, but the recorder does not like progress bars..)
[![asciicast](https://asciinema.org/a/hWGiJzJvJAJTxl0y6kljbrBd8.svg)](https://asciinema.org/a/hWGiJzJvJAJTxl0y6kljbrBd8)

Ideally i'd like an opinion on colors. On what curve to use for R_hat/acceptance (when should the value turn redd-ish) and what stuff we could add to the progress bar diplay. And if you like that or not. And what can we do for color-blind people.

Also, this should be tested with MPI and slurm, as i don't know if they will work nicely togethre (i think they should... but to be tested).

PhilipVinc commented:
@femtobit suggests that `R_hat > 1.1` starts to be _dangerous_ and therefore the color should turn orangeish/reddish around those values.

gcarleo commented:
I would say that more than redding things out on the acceptance, it would make more sense to red things out for high correlation times? ALPS had some coloring like this!

gpescia commented:
As discussed, I'm not sure if the coloring of the acceptance makes complete sense right now. 
In my application (MCMC with Gaussian kernel) an acceptance rate of 100% either means that my proposed step is much too small (correlation times will be huge) or the distribution is very flat. A very low acceptance rate usually indicates too large proposals (again big correlation times). So I am going for something between maybe 25% and 60%. 
I think that correlation times and acceptance rates are to some extend correlated and the color should be determined by some kind of score involving both values.

PhilipVinc commented:
make a proposal?

PhilipVinc committed:
make mc_stats repr print using colors


fix


color metropolis too


improve


progress bar


improve repr


black format


cleanup fx
